File: src/analysis/formats.ts
// 輸出 / 轉檔格式的**單一來源**：格式 union、能力表、編碼參數。
//
// 以前 "mp3" | "m4a" | "wav" 散在六個地方（render / api / exportPresets / tools / cli ×2）+ 三個對話框的 <Select>；
// 加一種格式要改七處，漏一處就是「選了 flac 默默出 mp3」（render.rs 的 `_ => mp3`）。
// 現在全部 import 這裡；Rust 端 formats.rs 是同一張表的鏡像，codec 字串用 golden 測試互相釘死。
export type RenderFormat = "mp3" | "m4a" | "wav" | "flac" | "ogg" | "opus" | "aiff";

export const RENDER_FORMATS: readonly RenderFormat[] = ["mp3", "m4a", "wav", "flac", "ogg", "opus", "aiff"];

export type BitDepth = 16 | 24 | 32;

export interface FormatCaps {
  id: RenderFormat;
  /** 顯示名稱（zh key）。 */
  label: string;
  /** ffmpeg 編碼器。 */
  codec: string;
  /** ffmpeg 容器（-f）。 */
  muxer: string;
  lossless: boolean;
  /** 章節只有 mp3（ID3 CHAP）與 m4a（QuickTime）寫得進去。 */
  chapters: boolean;
  /** 可選的位元深度（無損才有意義）。 */
  bitDepths: readonly BitDepth[];
  /** 一句話（zh key）。 */
  note: string;
}

export const FORMATS: Record<RenderFormat, FormatCaps> = {
  mp3: { id: "mp3", label: "MP3", codec: "libmp3lame", muxer: "mp3", lossless: false, chapters: true, bitDepths: [], note: "最通用；Podcast 平台都收" },
  m4a: { id: "m4a", label: "M4A（AAC）", codec: "aac", muxer: "mp4", lossless: false, chapters: true, bitDepths: [], note: "Apple 生態、同音質比 mp3 小" },
  wav: { id: "wav", label: "WAV", codec: "pcm_s16le", muxer: "wav", lossless: true, chapters: false, bitDepths: [16, 24, 32], note: "無損、檔案大；給後製用" },
  flac: { id: "flac", label: "FLAC", codec: "flac", muxer: "flac", lossless: true, chapters: false, bitDepths: [16, 24], note: "無損壓縮，約 wav 一半大" },
  ogg: { id: "ogg", label: "OGG（Vorbis）", codec: "libvorbis", muxer: "ogg", lossless: false, chapters: false, bitDepths: [], note: "開放格式；遊戲 / 網頁常用" },
  opus: { id: "opus", label: "Opus", codec: "libopus", muxer: "opus", lossless: false, chapters: false, bitDepths: [], note: "語音最省空間；瀏覽器都能播" },
  aiff: { id: "aiff", label: "AIFF", codec: "pcm_s16be", muxer: "aiff", lossless: true, chapters: false, bitDepths: [16, 24], note: "Mac 的無損；Logic / GarageBand 用" },
};

export function isRenderFormat(v: unknown): v is RenderFormat {
  return typeof v === "string" && (RENDER_FORMATS as readonly string[]).includes(v);
}

/** 副檔名 → 格式（不認得回 null）。 */
export function formatOfExt(ext: string): RenderFormat | null {
  const e = ext.replace(/^\./, "").toLowerCase();
  return isRenderFormat(e) ? e : null;
}

/** 這個格式的編碼參數（與 Rust formats.rs 的 codec_args 逐字相同，golden 測試釘死）。 */
export function codecArgs(format: RenderFormat, bitDepth?: BitDepth): string[] {
  switch (format) {
    case "mp3":
      return ["-c:a", "libmp3lame", "-q:a", "2"];
    case "m4a":
      return ["-c:a", "aac", "-b:a", "192k", "-movflags", "+faststart"];
    case "wav":
      return ["-c:a", bitDepth === 24 ? "pcm_s24le" : bitDepth === 32 ? "pcm_f32le" : "pcm_s16le"];
    case "flac":
      return ["-c:a", "flac", "-sample_fmt", bitDepth === 24 ? "s32" : "s16", "-compression_level", "8"];
    case "ogg":
      return ["-c:a", "libvorbis", "-q:a", "6"];
    case "opus":
      return ["-c:a", "libopus", "-b:a", "96k", "-vbr", "on"];
    case "aiff":
      return ["-c:a", bitDepth === 24 ? "pcm_s24be" : "pcm_s16be"];
  }
}

/** 輸出到這個格式時，哪些東西帶不過去（要列出來，不能默默少掉）。 */
export function droppedOnExport(format: RenderFormat, has: { chapters: boolean }): string[] {
  const out: string[] = [];
  if (has.chapters && !FORMATS[format].chapters) out.push("章節");
  return out;
}

/** 原檔的編碼器名稱（ffprobe codec_name）對到哪個格式；能對到就有機會直接複製不重編。 */
export function formatOfCodec(codec: string | null | undefined): RenderFormat | null {
  switch ((codec ?? "").toLowerCase()) {
    case "mp3":
      return "mp3";
    case "aac":
      return "m4a";
    case "flac":
      return "flac";
    case "vorbis":
      return "ogg";
    case "opus":
      return "opus";
    default:
      return null;
  }
}

File: src-tauri/src/formats.rs
//! 輸出 / 轉檔格式表：前端 `src/analysis/formats.ts` 的鏡像，codec 字串用兩邊的 golden 測試互相釘死。
//!
//! 以前 render.rs 的 `match plan.format { "wav" => …, "m4a" => …, _ => mp3 }`：不認得的格式**默默出 mp3**。
//! 現在不認得就是 `AppError::Invalid` —— 使用者選了 flac 就要拿到 flac，或拿到一個錯誤。

use crate::error::{AppError, AppResult};

pub const FORMATS: [&str; 7] = ["mp3", "m4a", "wav", "flac", "ogg", "opus", "aiff"];

#[allow(dead_code)]
pub fn is_format(f: &str) -> bool {
    FORMATS.contains(&f)
}

/// 章節只有 mp3（ID3 CHAP）與 m4a（QuickTime）寫得進去。
pub fn supports_chapters(format: &str) -> bool {
    matches!(format, "mp3" | "m4a")
}

#[allow(dead_code)]
pub fn is_lossless(format: &str) -> bool {
    matches!(format, "wav" | "flac" | "aiff")
}

/// (容器 -f, 編碼參數)。bit_depth 0 = 預設（16）。與 TS `codecArgs` 逐字相同。
pub fn codec_args(format: &str, bit_depth: u32) -> AppResult<(&'static str, Vec<String>)> {
    let s = |v: &[&str]| v.iter().map(|x| x.to_string()).collect::<Vec<_>>();
    Ok(match format {
        "mp3" => ("mp3", s(&["-c:a", "libmp3lame", "-q:a", "2"])),
        "m4a" => ("mp4", s(&["-c:a", "aac", "-b:a", "192k", "-movflags", "+faststart"])),
        "wav" => (
            "wav",
            s(&["-c:a", match bit_depth {
                24 => "pcm_s24le",
                32 => "pcm_f32le",
                _ => "pcm_s16le",
            }]),
        ),
        "flac" => ("flac", s(&["-c:a", "flac", "-sample_fmt", if bit_depth == 24 { "s32" } else { "s16" }, "-compression_level", "8"])),
        "ogg" => ("ogg", s(&["-c:a", "libvorbis", "-q:a", "6"])),
        "opus" => ("opus", s(&["-c:a", "libopus", "-b:a", "96k", "-vbr", "on"])),
        "aiff" => ("aiff", s(&["-c:a", if bit_depth == 24 { "pcm_s24be" } else { "pcm_s16be" }])),
        other => return Err(AppError::Invalid(format!("不支援的輸出格式：{other}（可用：{}）", FORMATS.join(" / ")))),
    })
}

/// ffprobe 的 codec_name 對到哪個格式（轉檔能不能直接 -c:a copy 用）。
pub fn format_of_codec(codec: &str) -> Option<&'static str> {
    match codec {
        "mp3" => Some("mp3"),
        "aac" => Some("m4a"),
        "flac" => Some("flac"),
        "vorbis" => Some("ogg"),
        "opus" => Some("opus"),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn golden_matches_ts_codec_args() {
        let j = |f: &str, b: u32| codec_args(f, b).unwrap().1.join(" ");
        assert_eq!(j("mp3", 0), "-c:a libmp3lame -q:a 2");
        assert_eq!(j("m4a", 0), "-c:a aac -b:a 192k -movflags +faststart");
        assert_eq!(j("wav", 0), "-c:a pcm_s16le");
        assert_eq!(j("wav", 24), "-c:a pcm_s24le");
        assert_eq!(j("wav", 32), "-c:a pcm_f32le");
        assert_eq!(j("flac", 0), "-c:a flac -sample_fmt s16 -compression_level 8");
        assert_eq!(j("flac", 24), "-c:a flac -sample_fmt s32 -compression_level 8");
        assert_eq!(j("ogg", 0), "-c:a libvorbis -q:a 6");
        assert_eq!(j("opus", 0), "-c:a libopus -b:a 96k -vbr on");
        assert_eq!(j("aiff", 0), "-c:a pcm_s16be");
        assert_eq!(j("aiff", 24), "-c:a pcm_s24be");
        assert_eq!(codec_args("m4a", 0).unwrap().0, "mp4");
        assert_eq!(codec_args("opus", 0).unwrap().0, "opus");
    }

    #[test]
    fn unknown_format_is_an_error_not_mp3() {
        let e = codec_args("wma", 0).unwrap_err();
        assert!(e.message().contains("wma"));
        assert!(!is_format("WAV"), "大小寫由前端正規化，這裡只認小寫");
    }

    #[test]
    fn chapters_only_in_mp3_and_m4a() {
        let with: Vec<&str> = FORMATS.iter().copied().filter(|f| supports_chapters(f)).collect();
        assert_eq!(with, vec!["mp3", "m4a"]);
        assert!(is_lossless("flac") && !is_lossless("opus"));
    }
}
